### PR TITLE
@std/json representation and serialization on js-runtime

### PR DESCRIPTION
**Pull request by dfellis**
This is not the final way I'd want to have to use `@std/json` -- you shouldn't have to import the nested types like it currently requires if you're only pulling in `toJSON`. I also hope to get the `json` interface done so you can automatically pull in a lot of related functions without a massive import line, but those can get fixed later.

### CODE REVIEW

File: compiler/src/index.ts [hunk 1]
```diff
@@ -1,5 +1,8 @@
 #!/usr/bin/env node
 
+// Make sure stack traces are *full*
```
Comment (dfellis) [inline compiler/src/index.ts:3]: Oh, I need to remove this.

File: compiler/src/lntoamm/opcodes.ts [hunk 2]
```diff
@@ -240,9 +241,24 @@ const addopcodes = (opcodes: object) => {
...
+                      const generics = [
+                        fulltypenameAst.get('opttypegenerics').get('generics').get('fulltypename').t
+                      ]
+                      fulltypenameAst.get('opttypegenerics').get('generics').get('cdr').getAll()
+                        .forEach(r => {
+                          generics.push(r.get('fulltypename').t)
+                        })
+                      return baseType.solidify(generics, scope)
+                    })() : scope.deepGet(arrayInnerTypeStr) as Type
+                  /*const arrayInnerType = scope.deepGet(
```
Comment (dfellis) [inline compiler/src/lntoamm/opcodes.ts:259]: Need to delete this line too

File: compiler/src/lntoamm/opcodes.ts [hunk 3]
```diff
@@ -421,9 +437,24 @@ const addopcodes = (opcodes: object) => {
...
+                        const generics = [
+                          fulltypenameAst.get('opttypegenerics').get('generics').get('fulltypename').t
+                        ]
+                        fulltypenameAst.get('opttypegenerics').get('generics').get('cdr').getAll()
+                          .forEach(r => {
+                            generics.push(r.get('fulltypename').t)
+                          })
+                        return baseType.solidify(generics, scope)
+                      })() : scope.deepGet(arrayInnerTypeStr) as Type
+                      /*const arrayInnerType = scope.deepGet(
```
Comment (dfellis) [inline compiler/src/lntoamm/opcodes.ts:455]: Same, note to delete this

File: std/json.ln [hunk 4]
```diff
@@ -0,0 +1,633 @@
+/* @std/json - A library for parsing, representing, and serializing to and from JSON
+ * This representation can handle every JSON type, and if you squint at it, it cannot represent an
```
Comment (depombo) [inline std/json.ln:2]: 😂

File: std/json.ln [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+ * invalid value. This is not entirely true without a small liberty being taken. If a JSONNode is an
+ * object, its children *should* be `KeyVal<string, JSONBase>` type, but if they're just `JSONBase`
+ * you could also just use the `toString` representation of the value as its key and the value as
+ * the value. Similarly, you should not ever have children attached to a `JSONBase`, but you can
+ * just ignore those children.
+ */
+export type IsObject {
+  isObject: bool,
+}
+// TODO: Fix issues with using `void` as a value in the compiler
```
Comment (depombo) [inline std/json.ln:12]: we should track this with a task if one doesn't exist

File: bdd/spec/032_json_spec.sh [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+Describe "JSON"
+  Describe "basic construction and printing"
+    before() {
+      sourceToAll "
+        from @std/app import start, print, exit
+        from @std/json import JSON, toJSON, toString, JSONBase, JSONNode, IsObject, Null
+
+        on start {
+          1.0.toJSON().print();
```
Comment (depombo) [inline bdd/spec/032_json_spec.sh:11]: This reads a bit odd, like a tuple in rust.

File: bdd/spec/032_json_spec.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+          true.toJSON().print();
+          'Hello, JSON!'.toJSON().print();
+          [1.0, 2.0, 5.0].toJSON().print();
+          toJSON().print();
```
Comment (depombo) [inline bdd/spec/032_json_spec.sh:15]: I really like that all base types implement this interface

**Review by depombo — APPROVED** [review]

File: bdd/spec/032_json_spec.sh (hunk 7 above)
Comment (dfellis) [reply bdd/spec/032_json_spec.sh:15]: *Almost* all base types. I haven't implemented it for the `int` types yet, though it should be trivial.

File: bdd/spec/032_json_spec.sh (hunk 6 above)
Comment (dfellis) [reply bdd/spec/032_json_spec.sh:11]: Oh, but it's the constant `1.0` having the `toJSON` method called on it. :)

That you've got it a bit confused with tuples is part why I haven't implemented tuples; keeping it unambiguous in the grammar is complex for multiple reasons (tuple definition, tuple type definition, and even tuple accessor syntax could be confusing).

File: std/json.ln (hunk 5 above)
Comment (dfellis) [reply std/json.ln:12]: I plan to make two tasks after this, at minimum. Not just `void` as value not working, but also the variable aliasing bug I ran across.

File: std/json.ln (hunk 4 above)
Comment (dfellis) [reply std/json.ln:2]: squint

File: bdd/spec/032_json_spec.sh (hunk 6 above)
Comment (depombo) [reply bdd/spec/032_json_spec.sh:11]: Yep, I agree. No tuples!

**Review by cdmistman — APPROVED** [review]

**Pull request merged**